src/foreign_key.c

```diff
@@ -0,0 +1,518 @@
+								  parentDelTrigger,
+								  parentUpdTrigger);
+	}
+}
+
+/*
+ * Copy foreign key constraints from the main table to a chunk.
+ */
+void
+ts_chunk_copy_referencing_fk(const Hypertable *ht, const Chunk *chunk)
+{
+	ListCell *lc;
+	List *chunks = list_make1((Chunk *) chunk);
+	List *fks = relation_get_referencing_fk(ht->main_table_relid);
+
+	Relation ht_rel = table_open(ht->main_table_relid, AccessShareLock);
+	foreach (lc, fks)
+	{
+		HeapTuple fk_tuple = lfirst(lc);
+		propagate_fk(ht_rel, fk_tuple, chunks);
+	}
+	table_close(ht_rel, NoLock);
+}
+
+void
+ts_fk_propagate(Constraint *c, Hypertable *ht)
+{
+	ListCell *lc1;
+	ListCell *lc2;
+
+	List *fks = relation_get_referencing_fk(ht->main_table_relid);
+	List *chunks = ts_chunk_get_by_hypertable_id(ht->fd.id);
+
+	Relation parentRel = table_open(ht->main_table_relid, AccessShareLock);
+
+	foreach (lc1, fks)
+	{
+		HeapTuple tuple = lfirst(lc1);
+		Form_pg_constraint fk = (Form_pg_constraint) GETSTRUCT(tuple);
+
+		int numfks;
+		AttrNumber conkey[INDEX_MAX_KEYS];
+		AttrNumber confkey[INDEX_MAX_KEYS];
+		Oid conpfeqop[INDEX_MAX_KEYS];
+		Oid conppeqop[INDEX_MAX_KEYS];
+		Oid conffeqop[INDEX_MAX_KEYS];
+#if PG15_GE
+		int numfkdelsetcols;
+		AttrNumber confdelsetcols[INDEX_MAX_KEYS];
+#endif
+
+		DeconstructFkConstraintRow(tuple,
+								   &numfks,
+								   conkey,
+								   confkey,
+								   conpfeqop,
+								   conppeqop,
+								   conffeqop
+#if PG15_GE
+								   ,
+								   &numfkdelsetcols,
+								   confdelsetcols
+#endif
+		);
+
+		Oid parentDelTrigger = InvalidOid;
+		Oid parentUpdTrigger = InvalidOid;
+
+		constraint_get_trigger(fk->oid, &parentUpdTrigger, &parentDelTrigger);
+
+		foreach (lc2, chunks)
+		{
+			Chunk *chunk = lfirst(lc2);
+
+			/*
+			 * We can skip the OSM chunks here because those
+			 * chunks are immutable and therefore don't need
+			 * the update and delete triggers.
+			 */
+			if (chunk->fd.osm_chunk)
+				continue;
+
+			clone_constraint_on_chunk(chunk,
+									  parentRel,
+									  fk,
+									  numfks,
+									  conkey,
+									  confkey,
+									  conpfeqop,
+									  conppeqop,
+									  conffeqop,
+#if PG15_GE
+									  numfkdelsetcols,
+									  confdelsetcols,
+#endif
+									  parentDelTrigger,
+									  parentUpdTrigger);
+		}
+	}
+
+	table_close(parentRel, AccessShareLock);
+}
+
+static void
+clone_constraint_on_chunk(const Chunk *chunk, Relation parentRel, Form_pg_constraint fk, int numfks,
+						  AttrNumber *conkey, AttrNumber *confkey, Oid *conpfeqop, Oid *conppeqop,
+						  Oid *conffeqop,
+#if PG15_GE
+						  int numfkdelsetcols, AttrNumber *confdelsetcols,
+#endif
+						  Oid parentDelTrigger, Oid parentUpdTrigger)
+{
+	AttrNumber mapped_confkey[INDEX_MAX_KEYS];
+	Relation pkrel = table_open(chunk->table_id, AccessShareLock);
+
+	// XXX FIXME allow non-primary as well and check if it's unique
+	// XXX FIXME bail when index is missing
+	Oid indexoid = RelationGetPrimaryKeyIndex(pkrel);
```

Comment:
Do you plan to fix this before merge or as a followup? The missing index case looks like it can break weirdly.

Reply:
Yes i will fix this as part of the PR